clique-grpc/src/lib.rs

```diff
@@ -1,18 +1,171 @@
+where
+    T: Service<Request> + Clone,
+    Request: Address,
+{
+    pub fn new() -> mpsc::UnboundedSender<(Request, oneshot::Sender<T::Response>)> {
+        let (tx, rx) = mpsc::unbounded();
+        tokio::spawn(FuturesRunner {
+            inner: FuturesUnordered::new(),
+            rx,
+        });
+
+        tx
+    }
 
-use clique::messaging::{Broadcast, RapidResponseFuture};
+    fn poll_enqueue(&mut self) {
+        self.rx
+            .for_each(|uri| {
+                let dst = Destination::try_from_uri(uri.clone()).unwrap();
+                let connector = util::Connector::new(HttpConnector::new(4));
+                let settings = client::Builder::new().http2_only(true).clone();
+                let mut make_client = client::Connect::with_builder(connector, settings);
 
-struct CliqueClient;
+                let client = make_client
+                    .make_service(dst) // we need to enqueue
+                    .map_err(|e| panic!("Unable to connect to destination"));
 
-impl Broadcast for CliqueClient {
-    fn broadcast(&self, req: RapidRequest) -> RapidResponseFuture {
-        unimplemented!()
+                self.inner.push(Box::new(client));
+                Ok(())
+            })
+            .poll();
     }
-    fn set_membership(&self, reciepients: Vec<Endpoint>) {
-        unimplemented!()
+}
+
+impl<Request, T> Future for FuturesRunner<Request, T>
+where
+    T: Service<Request> + Clone,
+    Request: Address,
+{
+    type Item = T::Response;
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Check if there are items to be run
+        self.poll_enqueue();
+
+        match self.inner.poll() {
+            Ok(Async::Ready(None)) => Ok(Async::NotReady), // There are no futures to run
+            Ok(Async::Ready(Some(val))) => Ok(Async::Ready(val)),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(e) => Err(e),
+        }
     }
-}
\ No newline at end of file
+}
+
+pub struct UniquePool<Request, T>
+where
+    T: Service<Request> + Clone,
+{
+    pool: HashMap<Uri, T::Response>,
+    buffering: usize,
+    max_buffering: usize,
+    enqueuer: mpsc::UnboundedSender<(Request, oneshot::Sender<T::Response>)>,
+}
+
+impl<Request, T> UniquePool<Request, T>
+where
+    T: Service<Request> + Clone,
+    Request: Address,
+{
+    pub fn new(max: usize) -> UniquePool<Request, T> {
+        UniquePool {
+            pool: HashMap::new(),
+            buffering: 0,
+            max_buffering: max,
+            enqueuer: FuturesRunner::new(),
+        }
+    }
+}
+
+trait Address {
+    type Address: Clone;
+    type Payload: Clone;
+
+    fn get_uri(&self) -> Self::Address;
+    fn get_payload(&self) -> Self::Payload;
+}
+
+impl<Request, T> Service<Request> for UniquePool<Request, T>
+where
+    T: Service<Request> + Clone,
+    Request: Address,
+{
+    type Response = T::Response;
+    type Error = (); // Need better error handling
```

Comment:
this can be `Box<std::error::Error + Send + Sync + 'static>` then in the bounds for this impl have a `T::Error: Into<Box<std::error::Error + Send + Sync + 'static>>`, this should make error handling a lot easier. Now all you need to do with a future si when you poll it to `map_err(Into::into)`